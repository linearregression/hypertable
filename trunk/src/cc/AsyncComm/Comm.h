/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef HYPERTABLE_COMMENGINE_H
#define HYPERTABLE_COMMENGINE_H

#include <string>

#include <boost/shared_ptr.hpp>
#include <boost/thread/mutex.hpp>
#include <boost/thread/xtime.hpp>

extern "C" {
#include <stdint.h>
}

#include "DispatchHandler.h"
#include "CommBuf.h"
#include "ConnectionHandlerFactory.h"
#include "HandlerMap.h"

using namespace std;

namespace hypertable {

  class Comm {

  public:

    Comm();

    ~Comm();

    int connect(struct sockaddr_in &addr, DispatchHandlerPtr &defaultHandlerPtr);

    int connect(struct sockaddr_in &addr, struct sockaddr_in &localAddr, DispatchHandlerPtr &defaultHandlerPtr);

    int listen(struct sockaddr_in &addr, ConnectionHandlerFactoryPtr &chfPtr);

    int listen(struct sockaddr_in &addr, ConnectionHandlerFactoryPtr &chfPtr, DispatchHandlerPtr &defaultHandlerPtr);

    int send_request(struct sockaddr_in &addr, time_t timeout, CommBufPtr &cbufPtr, DispatchHandler *responseHandler);

    int send_response(struct sockaddr_in &addr, CommBufPtr &cbufPtr);

    int get_local_address(struct sockaddr_in addr, struct sockaddr_in *localAddr);

    int create_datagram_receive_socket(struct sockaddr_in *addr, DispatchHandlerPtr &handlerPtr);

    int send_datagram(struct sockaddr_in &addr, struct sockaddr_in &sendAddr, CommBufPtr &cbufPtr);

    int set_timer(uint64_t durationMillis, DispatchHandler *handler);

    int set_timer_absolute(boost::xtime expireTime, DispatchHandler *handler);

    int close_socket(struct sockaddr_in &addr);

  private:

    int connect_socket(int sd, struct sockaddr_in &addr, DispatchHandlerPtr &defaultHandlerPtr);

    boost::mutex  m_mutex;
    std::string   m_app_name;
    HandlerMap    m_handler_map;
    Reactor      *m_timer_reactor;
  };

  typedef boost::shared_ptr<Comm> CommPtr;

}

#endif // HYPERTABLE_COMMENGINE_H
