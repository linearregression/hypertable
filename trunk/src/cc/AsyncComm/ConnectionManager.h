/**
 * Copyright 2007 Doug Judd (Zvents, Inc.)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at 
 *
 * http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#ifndef HYPERTABLE_CONNECTIONMANAGER_H
#define HYPERTABLE_CONNECTIONMANAGER_H

#include <string>

#include <boost/thread/condition.hpp>
#include <boost/thread/mutex.hpp>
#include <boost/thread/thread.hpp>

extern "C" {
#include <time.h>
#include <sys/time.h>
}

#include "CallbackHandler.h"

namespace hypertable {

  class Comm;
  class Event;

  class ConnectionManager {

  public:

    static const int RETRY_INTERVAL = 10;

    ConnectionManager(std::string &waitMessage) : mWaitMessage(waitMessage) { return; }

    ConnectionManager(const char *waitMessage) : mWaitMessage(waitMessage) { return; }

    void Initiate(Comm *comm, struct sockaddr_in &addr, time_t timeout) {
      mHandler = new Callback(comm, addr, timeout);
      mThread = new boost::thread(*this);
      return;
    }

    bool WaitForConnection(long maxWaitSecs) {
      struct timeval tval;
      long elapsed = 0;
      long starttime;

      gettimeofday(&tval, 0);
      starttime = tval.tv_sec;

      while (elapsed < maxWaitSecs) {

	if (mHandler->WaitForEvent(maxWaitSecs-elapsed))
	  return true;

	gettimeofday(&tval, 0);
	elapsed = tval.tv_sec - starttime;
      }
      return false;
    }

    void operator()();

    class Callback : public CallbackHandler {
    public:
      Callback(Comm *comm, struct sockaddr_in &addr, time_t timeout) 
	: mComm(comm), mAddr(addr), mTimeout(timeout), mConnected(false), mMutex(), mCond() { return; }
      virtual ~Callback() { return; }
      virtual void handle(Event &event);
      void SendConnectRequest();
      bool WaitForEvent(long maxWaitSecs);
      bool WaitForEvent();

    private:
      Comm               *mComm;
      struct sockaddr_in  mAddr;
      time_t              mTimeout;
      bool                mConnected;
      boost::mutex        mMutex;
      boost::condition    mCond;
    };

  private:
    Callback       *mHandler;
    boost::thread  *mThread;
    std::string     mWaitMessage;
  };

}


#endif // HYPERTABLE_CONNECTIONMANAGER_H
