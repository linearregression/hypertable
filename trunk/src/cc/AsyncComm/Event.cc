/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */


#include <iostream>
using namespace std;

extern "C" {
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
}

#include "Common/Error.h"
#include "Common/StringExt.h"

#include "Event.h"

namespace hypertable {

  std::string Event::toString() {
    string dstr;

    dstr = "Event: type=";
    if (type == CONNECTION_ESTABLISHED)
      dstr += "CONNECTION_ESTABLISHED";
    else if (type == DISCONNECT)
      dstr += "DISCONNECT";
    else if (type == MESSAGE) {
      dstr += "MESSAGE protocol=";
      if (header->protocol < Header::PROTOCOL_MAX)
	dstr += Header::protocolStrings[header->protocol];
      else
	dstr += "unknown";
      dstr += " id=" + header->id;
      dstr += " threadGroup=" + header->threadGroup;
      dstr += " headerLen=" + header->headerLen;
      dstr += " totalLen=" + header->totalLen;
    }
    else if (type == ERROR)
      dstr += "ERROR";
    else
      dstr += (int)type;

    if (error != Error::OK)
      dstr += (std::string)" \"" + Error::GetText(error) + "\"";

    dstr += " from=";
    dstr += (std::string)inet_ntoa(addr.sin_addr) + ":" + (int)ntohs(addr.sin_port);

    return dstr;
  }

}


