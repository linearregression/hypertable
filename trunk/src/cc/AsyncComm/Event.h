/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#ifndef HYPERTABLE_EVENT_H
#define HYPERTABLE_EVENT_H

#include <iostream>
#include <string>

#include <boost/shared_ptr.hpp>

extern "C" {
#include <stdint.h>
#include <netinet/in.h>
}

using namespace std;

#include "Message.h"

namespace hypertable {

  class Event {

  public:

    enum Type { CONNECTION_ESTABLISHED, DISCONNECT, MESSAGE, ERROR };

    Event(Type ct, struct sockaddr_in &a, int err=0, Message::HeaderT *h=0) 
      : type(ct), addr(a), error(err), header(h) {
      if (h != 0) {
	message = ((uint8_t *)header) + header->headerLen;
	messageLen = header->totalLen - header->headerLen;
      }
      else {
	message = 0;
	messageLen = 0;
      }
    }

    ~Event() {
      delete [] header;
    }

    Type                type;
    struct sockaddr_in  addr;
    int                 error;
    Message::HeaderT   *header;
    uint8_t            *message;
    size_t              messageLen;

    std::string toString();
    void Display() { cerr << toString() << endl; }
  };

  typedef boost::shared_ptr<Event> EventPtr;

}

#endif // HYPERTABLE_EVENT_H
