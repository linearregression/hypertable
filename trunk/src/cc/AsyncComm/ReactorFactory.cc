/**
 * Copyright 2007 Doug Judd (Zvents, Inc.)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at 
 *
 * http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "ReactorFactory.h"
#include "ReactorRunner.h"
using namespace hypertable;

#include <cassert>

extern "C" {
#include <signal.h>
}

vector<Reactor *> ReactorFactory::msReactors;
atomic_t ReactorFactory::msNextReactor = ATOMIC_INIT(0);



/**
 * Method to initialize the reactor factory
 *
 * @param reactorCount number of reactors to initialize
 */
void ReactorFactory::Initialize(uint16_t reactorCount) {
  Reactor *reactor;
  ReactorRunner rrunner;
  signal(SIGPIPE, SIG_IGN);
  assert(reactorCount > 0);
  for (uint16_t i=0; i<reactorCount; i++) {
    reactor = new Reactor();
    msReactors.push_back(reactor);
    rrunner.SetReactor(reactor);
    reactor->threadPtr = new boost::thread(rrunner);
  }
}

