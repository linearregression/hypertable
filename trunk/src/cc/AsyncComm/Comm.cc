/**
 * Copyright 2007 Doug Judd (Zvents, Inc.)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at 
 *
 * http://www.apache.org/licenses/LICENSE-2.0 
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


//#define DISABLE_LOG_DEBUG

#include <cassert>
#include <iostream>
using namespace std;

extern "C" {
#if defined(__APPLE__)
#include <arpa/inet.h>
#endif
#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <poll.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <arpa/inet.h>
}

#include "Common/Error.h"
#include "Common/FileUtils.h"
#include "Common/System.h"

#include "ReactorFactory.h"
#include "Comm.h"
#include "IOHandlerAccept.h"
#include "IOHandlerData.h"
using namespace hypertable;


/**
 */
Comm::Comm(uint32_t handlerCount) {

  if (ReactorFactory::msReactors.size() == 0) {
    LOG_ERROR("ReactorFactory::Initialize must be called before creating AsyncComm::Comm object");
    DUMP_CORE;
  }
  // TODO: Implement me ...
  assert(handlerCount == 0);
  mEventQueue = 0;
}



/**
 */
int Comm::Connect(struct sockaddr_in &addr, time_t timeout, CallbackHandler *defaultHandler) {
  int sd;
  IOHandlerDataPtr dataHandlerPtr;

  if (mConnMap.Lookup(addr, dataHandlerPtr))
    return Error::COMM_ALREADY_CONNECTED;

  if ((sd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
    cerr << "socket() failure: " << strerror(errno) << endl;
    exit(1);
  }

  // Set to non-blocking
  FileUtils::SetFlags(sd, O_NONBLOCK);

#if defined(__APPLE__)
  int one = 1;
  if (setsockopt(sd, SOL_SOCKET, SO_NOSIGPIPE, &one, sizeof(one)) < 0)
    LOG_VA_WARN("setsockopt(SO_NOSIGPIPE) failure: %s", strerror(errno));
#endif

  dataHandlerPtr.reset( new IOHandlerData(sd, addr, defaultHandler, mConnMap, mEventQueue) );
  mConnMap.Insert(dataHandlerPtr);
  dataHandlerPtr->SetTimeout(timeout);

  while (connect(sd, (struct sockaddr *)&addr, sizeof(struct sockaddr_in)) < 0) {
    if (errno == EINTR) {
      poll(0, 0, 1000);
      continue;
    }
    else if (errno == EINPROGRESS) {
      dataHandlerPtr->StartPolling();
      dataHandlerPtr->AddPollInterest(Reactor::READ_READY|Reactor::WRITE_READY);
      return Error::OK;
    }
    LOG_VA_ERROR("connect() failure : %s", strerror(errno));
    exit(1);
  }

  dataHandlerPtr->StartPolling();

  return Error::OK;
}


/**
 *
 */
int Comm::Listen(uint16_t port, ConnectionHandlerFactory *hfactory, CallbackHandler *defaultHandler) {
  struct sockaddr_in addr;
  int one = 1;
  int sd;

  port = htons(port);

  if ((sd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
    cerr << "socket() failure: " << strerror(errno) << endl;
    exit(1);
  }

#if defined(__linux__)
  if (setsockopt(sd, SOL_TCP, TCP_NODELAY, &one, sizeof(one)) < 0)
    cerr << "setsockopt(TCP_NODELAY) failure: " << strerror(errno) << endl;
#elif defined(__APPLE__)
  if (setsockopt(sd, SOL_SOCKET, SO_NOSIGPIPE, &one, sizeof(one)) < 0)
    LOG_VA_WARN("setsockopt(SO_NOSIGPIPE) failure: %s", strerror(errno));
#endif

  if (setsockopt(sd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one)) < 0)
    cerr << "setsockopt(SO_REUSEADDR) failure: " << strerror(errno) << endl;

  // create address structure to bind to - any available port - any address
  memset(&addr, 0 , sizeof(sockaddr_in));
  addr.sin_family = AF_INET;
  addr.sin_addr.s_addr = htonl(INADDR_ANY);
  addr.sin_port = port;

  // bind socket
  if ((bind(sd, (const sockaddr *)&addr, sizeof(sockaddr_in))) < 0) {
    cerr << "bind() failure: " << strerror(errno) << endl;
    exit(1);
  }

  if (listen(sd, 64) < 0) {
    cerr << "listen() failure: " << strerror(errno) << endl;
    exit(1);
  }

  IOHandlerAccept *acceptHandler = new IOHandlerAccept(sd, defaultHandler, mConnMap, mEventQueue, hfactory);
  acceptHandler->StartPolling();

  return 0;
}


int Comm::SendRequest(struct sockaddr_in &addr, CommBuf *cbuf, CallbackHandler *responseHandler) {
  boost::mutex::scoped_lock lock(mMutex);
  IOHandlerDataPtr dataHandlerPtr;
  Message::HeaderT *mheader = (Message::HeaderT *)cbuf->data;
  int error = Error::OK;

  if (!mConnMap.Lookup(addr, dataHandlerPtr)) {
    LOG_VA_ERROR("No connection for %s:%d", inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));
    return Error::COMM_NOT_CONNECTED;
  }

  mheader->flags |= Message::FLAGS_MASK_REQUEST;

  if ((error = dataHandlerPtr->SendMessage(cbuf, responseHandler)) != Error::OK)
    dataHandlerPtr->Shutdown();

  return error;
}


int Comm::SendResponse(struct sockaddr_in &addr, CommBuf *cbuf) {
  boost::mutex::scoped_lock lock(mMutex);
  IOHandlerDataPtr dataHandlerPtr;
  Message::HeaderT *mheader = (Message::HeaderT *)cbuf->data;
  int error = Error::OK;

  if (!mConnMap.Lookup(addr, dataHandlerPtr)) {
    LOG_VA_ERROR("No connection for %s:%d", inet_ntoa(addr.sin_addr), ntohs(addr.sin_port));
    return Error::COMM_NOT_CONNECTED;
  }

  mheader->flags &= Message::FLAGS_MASK_RESPONSE;

  if ((error = dataHandlerPtr->SendMessage(cbuf)) != Error::OK)
    dataHandlerPtr->Shutdown();

  return error;
}
