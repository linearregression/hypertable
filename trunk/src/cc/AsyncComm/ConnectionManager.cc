/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <iostream>

extern "C" {
#include <poll.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
}

#include "Common/Error.h"
#include "Common/Logger.h"

#include "AsyncComm/Comm.h"

#include "ConnectionManager.h"

using namespace hypertable;
using namespace std;


/**
 * This method adds a connection to the connection manager if it is not already added.
 * It first allocates a ConnectionStateT structure for the connection and initializes
 * it to the non-connected state.  It then attempts to establish the connection.
 *
 * @param addr         The address to connect to
 * @param timeout      When connection dies, wait this many seconds before attempting to reestablish
 * @param serviceName  The name of the service we're connnecting to (used for better log messages)
 * @param handler      This is the default handler to install on the connection.  All events get changed through to this handler.
 */
void ConnectionManager::Add(struct sockaddr_in &addr, time_t timeout, const char *serviceName, DispatchHandler *handler) {
  boost::mutex::scoped_lock lock(mImpl->mutex);
  SockAddrMapT<ConnectionStatePtr> iter;
  ConnectionState *connState;
  int error;

  if (mImpl->connMap.find(addr) != mImpl->connMap.end())
    return;

  connState = new ConnectionState();
  connState->connected = false;
  connState->addr = addr;
  memset(&connState->localAddr, 0, sizeof(struct sockaddr_in));
  connState->timeout = timeout;
  connState->handler = handler;
  connState->serviceName = (serviceName) ? serviceName : "";
  boost::xtime_get(&connState->nextRetry, boost::TIME_UTC);

  mImpl->connMap[addr] = ConnectionStatePtr(connState);

  {
    boost::mutex::scoped_lock connLock(connState->mutex);
    SendConnectRequest(connState);
  }
}


/**
 * This method adds a connection to the connection manager if it is not already added.
 * It first allocates a ConnectionStateT structure for the connection and initializes
 * it to the non-connected state.  It then attempts to establish the connection.
 *
 * @param addr         The address to connect to
 * @param localAddr    The local address to bind to
 * @param timeout      When connection dies, wait this many seconds before attempting to reestablish
 * @param serviceName  The name of the service we're connnecting to (used for better log messages)
 * @param handler      This is the default handler to install on the connection.  All events get changed through to this handler.
 */
void ConnectionManager::Add(struct sockaddr_in &addr, struct sockaddr_in &localAddr, time_t timeout, const char *serviceName, DispatchHandler *handler) {
  boost::mutex::scoped_lock lock(mImpl->mutex);
  SockAddrMapT<ConnectionStatePtr> iter;
  ConnectionState *connState;
  int error;

  if (mImpl->connMap.find(addr) != mImpl->connMap.end())
    return;

  connState = new ConnectionState();
  connState->connected = false;
  connState->addr = addr;
  connState->localAddr = localAddr;
  connState->timeout = timeout;
  connState->handler = handler;
  connState->serviceName = (serviceName) ? serviceName : "";
  boost::xtime_get(&connState->nextRetry, boost::TIME_UTC);

  mImpl->connMap[addr] = ConnectionStatePtr(connState);

  {
    boost::mutex::scoped_lock connLock(connState->mutex);
    SendConnectRequest(connState);
  }
}



bool ConnectionManager::WaitForConnection(struct sockaddr_in &addr, long maxWaitSecs) {
  ConnectionStatePtr connStatePtr;

  {
    boost::mutex::scoped_lock lock(mImpl->mutex);
    SockAddrMapT<ConnectionStatePtr>::iterator iter = mImpl->connMap.find(addr);
    assert(iter != mImpl->connMap.end());
    connStatePtr = (*iter).second;
  }

  {
    boost::mutex::scoped_lock connLock(connStatePtr->mutex);
    boost::xtime dropTime, now;

    boost::xtime_get(&dropTime, boost::TIME_UTC);
    dropTime.sec += maxWaitSecs;

    while (!connStatePtr->connected) {
      connStatePtr->cond.timed_wait(connLock, dropTime);
      boost::xtime_get(&now, boost::TIME_UTC);
      if (xtime_cmp(now, dropTime) >= 0)
	return false;
    }
  }

  return true;
}



/**
 * Attempts to establish a connection for the given ConnectionState object.  If a failure
 * occurs, it prints an error message and then schedules a retry by updating the
 * nextRetry member of the connState object and pushing it onto the retry heap
 *
 * @param connState The connection state record
 */
void ConnectionManager::SendConnectRequest(ConnectionState *connState) {
  int error;

  if (connState->localAddr.sin_port != 0)
    error = mImpl->comm->Connect(connState->addr, connState->localAddr, this);
  else
    error = mImpl->comm->Connect(connState->addr, this);

  if (error == Error::COMM_ALREADY_CONNECTED) {
    connState->connected = true;
    connState->cond.notify_all();
  }
  else if (error != Error::OK) {
    if (connState->serviceName != "") {
      LOG_VA_ERROR("Connection attempt to %s at %s:%d failed - %s.  Will retry again in %d seconds...", 
		   connState->serviceName.c_str(), inet_ntoa(connState->addr.sin_addr), ntohs(connState->addr.sin_port),
		   Error::GetText(error), connState->timeout);
    }
    else {
      LOG_VA_ERROR("Connection attempt to service at %s:%d failed - %s.  Will retry again in %d seconds...",
		   inet_ntoa(connState->addr.sin_addr), ntohs(connState->addr.sin_port),
		   Error::GetText(error), connState->timeout);
    }

    // reschedule
    boost::xtime_get(&connState->nextRetry, boost::TIME_UTC);
    connState->nextRetry.sec += connState->timeout;

    // add to retry heap
    mImpl->retryQueue.push(connState);
    mImpl->retryCond.notify_one();
  }

}



/**
 *
 */
int ConnectionManager::Remove(struct sockaddr_in &addr) {
  bool doClose = false;
  int error = Error::OK;

  {
    boost::mutex::scoped_lock lock(mImpl->mutex);
    SockAddrMapT<ConnectionStatePtr>::iterator iter = mImpl->connMap.find(addr);

    if (iter != mImpl->connMap.end()) {
      boost::mutex::scoped_lock connLock((*iter).second->mutex);
      if ((*iter).second->connected)
	doClose = true;
      else
	(*iter).second->connected = true;  // will prevent further connection attempts
      mImpl->connMap.erase(iter);
    }
  }

  if (doClose)
    error = mImpl->comm->CloseSocket(addr);

  return error;
}



/**
 * This is the AsyncComm dispatch handler method.  It gets called for each connection
 * related event (establishment, disconnect, etc.) for each connection.  For connect
 * events, the connection's connected flag is set to true and it's condition variable
 * is signaled.  For all other events (e.g. disconnect or error), the connection's
 * connected flag is set to false and a retry is scheduled.
 *
 * @param eventPtr shared pointer to event object
 */
void ConnectionManager::handle(EventPtr &eventPtr) {
  boost::mutex::scoped_lock lock(mImpl->mutex);
  SockAddrMapT<ConnectionStatePtr>::iterator iter = mImpl->connMap.find(eventPtr->addr);

  if (iter != mImpl->connMap.end()) {
    boost::mutex::scoped_lock connLock((*iter).second->mutex);
    ConnectionState *connState = (*iter).second.get();

    if (eventPtr->type == Event::CONNECTION_ESTABLISHED) {
      connState->connected = true;
      connState->cond.notify_all();
    }
    else {
      if (!mImpl->quietMode) {
	LOG_VA_INFO("%s; Problem connecting to %s, will retry in %d seconds...",
		    eventPtr->toString().c_str(), connState->serviceName.c_str(), connState->timeout);
      }
      connState->connected = false;
      // this logic could proably be smarter.  For example, if the last
      // connection attempt was a long time ago, then schedule immediately
      // otherwise, if this event is the result of an immediately prior connect
      // attempt, then do the following
      boost::xtime_get(&connState->nextRetry, boost::TIME_UTC);
      connState->nextRetry.sec += connState->timeout;
      // add to retry heap
      mImpl->retryQueue.push(connState);
      mImpl->retryCond.notify_one();
    }
    
    // Chain event to application supplied handler
    if (connState->handler)
      connState->handler->handle(eventPtr);
  }
  else {
    LOG_VA_WARN("Unable to find connection for %s:%d in map.", inet_ntoa(eventPtr->addr.sin_addr), ntohs(eventPtr->addr.sin_port));
  }
}



/**
 * This is the boost::thread run method.
 */
void ConnectionManager::operator()() {
  boost::mutex::scoped_lock lock(mImpl->mutex);
  ConnectionStatePtr connStatePtr;

  while (true) {

    while (mImpl->retryQueue.empty())
      mImpl->retryCond.wait(lock);

    connStatePtr = mImpl->retryQueue.top();

    if (!connStatePtr->connected) {
      {
	boost::mutex::scoped_lock connLock(connStatePtr->mutex);
	boost::xtime now;
	boost::xtime_get(&now, boost::TIME_UTC);

	if (xtime_cmp(connStatePtr->nextRetry, now) <= 0) {
	  mImpl->retryQueue.pop();
	  /**
	     if (!mImpl->quietMode) {
	     LOG_VA_INFO("Attempting to re-establish connection to %s at %s:%d...",
	     connStatePtr->serviceName.c_str(), inet_ntoa(connStatePtr->addr.sin_addr), ntohs(connStatePtr->addr.sin_port));
	     }
	  */
	  SendConnectRequest(connStatePtr.get());
	  return;
	}
      }
      mImpl->retryCond.timed_wait(lock, connStatePtr->nextRetry);
    }
    else
      mImpl->retryQueue.pop();
  }
}


