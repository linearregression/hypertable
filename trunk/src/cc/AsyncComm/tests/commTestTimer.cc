/**
 * Copyright (C) 2007 Doug Judd (Zvents, Inc.)
 * 
 * This file is part of Hypertable.
 * 
 * Hypertable is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or any later version.
 * 
 * Hypertable is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

#include <cstdlib>

extern "C" {
#include <poll.h>
#include <signal.h>
#include <sys/types.h>
#include <unistd.h>
}

#include <boost/thread/thread.hpp>

#include "Common/Error.h"
#include "Common/FileUtils.h"
#include "Common/Logger.h"
#include "Common/TestHarness.h"
#include "Common/System.h"
#include "Common/Usage.h"

#include "AsyncComm/Comm.h"
#include "AsyncComm/ConnectionManager.h"
#include "AsyncComm/Event.h"
#include "AsyncComm/ReactorFactory.h"

#include "CommTestThreadFunction.h"

using namespace hypertable;

namespace {
  const char *usage[] = {
    "usage: commTestTimer",
    "",
    "This program tests tests timer logic of AsyncComm.",
    0
  };

  /**
   * 
   */
  class TimerHandler : public DispatchHandler {
  public:
    TimerHandler(const char *msg, ostream &out) : mMsg(msg), mOut(out) { return; }
    virtual void handle(EventPtr &eventPtr) {
      mOut << eventPtr->toString().c_str() << " (" << mMsg << ")" << endl;
    }
  private:
    const char *mMsg;
    ostream    &mOut;
  };

  struct TimerRec {
    uint32_t delay;
    const char *msg;
  };

  struct TimerRec history[] = {
    { 1, "Local" },
    { 2, "media" },
    { 3, "have" },
    { 4, "reported" },
    { 5, "eyewitness" },
    { 6, "accounts" },
    { 7, "of" },
    { 8, "a" },
    { 9, "fiery" },
    { 10, "ball" },
    { 35, 0 }
  };
}


int main(int argc, char **argv) {
  Comm *comm;
  int error;
  TestHarness harness("commTestTimer");
  bool golden = false;
  TimerHandler *timerHandler;
  int waitTime = 0;
  ostream &out = harness.GetLogStream();

  if (argc > 1) {
    if (!strcmp(argv[1], "--golden"))
      golden = true;
    else
      Usage::DumpAndExit(usage);
  }

  srand(8876);

  System::Initialize(argv[0]);
  ReactorFactory::Initialize(1);

  comm = new Comm();

  for (int i=0; history[i].msg; i++) {
    timerHandler = new TimerHandler(history[i].msg, out);
    if ((error = comm->SetTimer(history[i].delay*1000, timerHandler)) != Error::OK) {
      LOG_VA_ERROR("Problem setting timer - %s", Error::GetText(error));
      exit(1);
    }
    if (history[i].delay > waitTime)
      waitTime = history[i].delay;
  }

  poll(0, 0, (waitTime+1)*1000);

  if (!golden)
    harness.ValidateAndExit("commTestTimer.golden");

  harness.RegenerateGoldenFile("commTestTimer.golden");

  delete comm;

  return 0;
}
