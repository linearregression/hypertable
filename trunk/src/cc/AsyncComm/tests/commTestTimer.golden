Event: type=TIMER from=0.0.0.0:0 (Local)
Event: type=TIMER from=0.0.0.0:0 (media)
Event: type=TIMER from=0.0.0.0:0 (have)
Event: type=TIMER from=0.0.0.0:0 (reported)
Event: type=TIMER from=0.0.0.0:0 (eyewitness)
Event: type=TIMER from=0.0.0.0:0 (accounts)
Event: type=TIMER from=0.0.0.0:0 (of)
Event: type=TIMER from=0.0.0.0:0 (a)
Event: type=TIMER from=0.0.0.0:0 (fiery)
Event: type=TIMER from=0.0.0.0:0 (ball)
